tortoise/tortoise.go

```diff
@@ -408,41 +407,46 @@ func (t *turtle) getFullVote(ctx context.Context, block *blockInfo) (sign, voteR
 	return against, reasonCoinflip, nil
 }
 
-// onLayerTerminated is expected to be called when hare terminated for a layer.
-// Internally tortoise will verify all layers before the last one if there are no gaps in
-// terminated layers.
-func (t *turtle) onLayerTerminated(ctx context.Context, lid types.LayerID) error {
-	t.logger.With().Debug("on layer terminated", lid)
+func (t *turtle) onLayer(ctx context.Context, lid types.LayerID) error {
+	t.logger.With().Debug("on layer", lid)
 	defer t.evict(ctx)
 	if err := t.updateLayer(t.logger, lid); err != nil {
```

Comment:
this is where the reference height is calculated. isn't it still possible that onLayer() is called after the first block of the epoch is fetched?

Reply:
updateRefHeight for block in layer X is called only after layer X was processed

for example if tortoise receives:

on_block bid=AA lid=100. if processed is 99 update ref height is not triggered, because it may fail
on_layer lid=100. processed is set to 100, process also looks up what blocks are in the internal state and triggers update ref height for them
on_block bid=BB lid=100. processed is 100, so now failure is not expected and update ref height is triggered immediately